_posts/2016-09-14-programming-model.md

```diff
@@ -0,0 +1,128 @@
+
+With GFX, you pass everything into the draw call:
+{% highlight Rust %}
+encoder.draw(
+	&slice, // contains the range of vertices and (optionally) an index buffer pointer
+	&pso,	// the pipeline state object, having all the graphics state encapsulated
+	&data	// the pipeline data, such as: vertex buffers, constant buffers, textures, etc
+);
+{% endhighlight %}
+
+Note that you don't pass *all* the state in each draw call, but only what is needed. We encode as much dependency information as possible into the type system, while still allowing one to trade the run-time correctness for the compile-time in case where more flexibility is needed.
+
+
+## Command Buffers
+
+Last-gen APIs (such as OpenGL and Dx11) suffered from the limited multi-threading capabilities, because you could only render in one thread that owned the graphics context. The CPU rendering overhead thus would often become the bottleneck of applications, reducing the number of draw calls sent through the pipeline.
+
+This issue has been properly addressed by all of the current-gen APIs (Dx12, Vulkan, etc) - they introduced the concept of a command buffer that can be filled up independently of the rendering context. Conceptually, a command buffer is an opaque object that is optimized for direct execution on your graphics hardware (it's not portable, much like the machine code). The idea is - you can populate multiple command buffers on multiple threads simultaneously, and then only "submit" them to the single graphics queue that would execute them on the hardware. Obviously, the concept complicates programming a bit. In exchange, it allows multi-threading your renderer and thus drawing many more objects on screen.
+
+GFX functions the same way. We have the abstract command buffer trait implemented by the backends, and we wrap them into `gfx::Encoder` objects, which have methods for drawing as well as updating the buffers/textures. Once you are ready to execute the encoded commands, you need to pass it to `gfx::Device` for submission. If the commands are encoded on a different thread, you may use channels or sharing to coordinate the submission.
+
+For an example of some multi-threading, you can check out [yasteroids](https://github.com/kvark/yasteroids). In short, the context-owning thread does this:
+{% highlight Rust %}
+let mut encoder = dev_recv.recv().unwrap(); // receive the next command buffer
+encoder.flush(&mut device); // submit for execution, also clears the buffer
+dev_send.send(encoder).unwrap(); // send it back for reuse
+window.swap_buffers().unwrap(); // flip the front and back buffers, showing the frame
+device.cleanup(); // delete the resources no longer used
+{% endhighlight %}
```

Comment:
> `dev_recv.recv().unwrap();`

Clarify that it's received from a channel in comments

Comments were aligned on the other example.